### ISSUE TITLE
Revising langUsage

### ISSUE THREAD

**Issue by nathangibson** [srophe/srophe-app-data#663]
I had a chance to discuss this yesterday with @dlschwartz and @davidamichelson, but not with @jsaintlauren or @tacarlson. Also @wsalesky your comments on the technical feasibility/advisability of this are welcome. 

We have been struggling for a while with how to best document the language codes we use. A while back, @tacarlson wrote up this helpful documentation page http://syriaca.org/exist/apps/srophe/documentation/isostandards.html and in any case we need a narrative like this. However, we also need a machinable way for scripts (whether by @wsalesky or external partners) to grab our descriptions/usage guidelines of specific codes. We should also declare within each TEI document what we mean by the language codes we use in that document. 

So far we have used `<langUsage>` in the profileDesc section of the header as a kind of all-purpose solution by putting there the description of *all* of the language codes that *might* conceivably be used in that document. There are some problems with this. First, langUsage is supposed to list the languages used "within a text" or "within a document" (http://www.tei-c.org/release/doc/tei-p5-doc/en/html/HD.html#HD41). By including a bunch of languages here, we are claiming each document uses languages that it doesn't. Second, we have not succeeded in keeping these lists uniform, and we won't unless we have a global list.

In order to document our language code usage in a way that @wsalesky can grab, I've collected all the language tags I could find listed on the documentation page and in the langUsage elements and put them in a draft of our ODD file as a value list for `@xml:lang`: https://github.com/srophe/srophe-eXist-app/blob/dev/srophe-app/documentation/syriaca-tei-main.odd#L46. Scripts can search on the `@ident` and pull either the `<gloss>` (language name) or `<desc>` description of how we use it. An additional advantage is that if you are editing a doc in oXygen and are running validation from the RNC file generated by this ODD, the particular language codes we use will be automatically suggested and you will see our own description of them. 

Here are my recommendations:
 1. That the list in this ODD file should be our master machine-readable list of language codes and descriptions. 
 2. That we update the langUsage element in all our records to contain *only* the languages actually used in that file and to use the master descriptions for those languages. This could be a script we run now as well as a maintenance script that we run periodically or on upload/edit in case we have added data with new language codes to a record.

Responses? Objections?

**Comment by dlschwartz** [comment]
@nathangibson I think 2. is a good idea as is. I have a slight concern about 1. as stated. SPEAR has a separate ODD and other modules designed in the future may as well. I think we should consider a remote list that each ODD calls to for this purpose. It should work the same when producing TEI but would give the added benefit that we would only have one list to maintain for as many ODDs as we have in play. 

I need to look into how this works in detail. I am wondering if the TEI file from which the HTML of the documentation page is made couldn't be encoded in such a way as to handle this task. That way one thing would be our canonical metadata on languages, the basis for our HTML documentation page on languages, and the source for attribute values (with glosses and descriptions) in our ODD. 

For the time being, I think keeping a list in this ODD is fine. Let me check into the details of calling such a list in an ODD and I'll get back to you.

**Comment by tacarlson** [comment]
I agree that it's a problem to have the header asserting languages are used when they are not, but I wonder why we need machine-parsable language use documentation at all, or even a `<langUsage>` in the header.  What is the anticipated use of these affordances which is so compelling?  The easiest solution for the use cases which appear to me is to maintain good prose documentation and strip out the `<langUsage>` elements in the header so that they do not get out-of-date.

________________________________

**Comment by nathangibson** [comment]
@dlschwartz Thanks, Dan, and I'll look forward to hearing what you find out about pulling such values into the ODD file. Perhaps ODD has some built-in mechanism for this. xi:include can do this for XML files in general, but when we discussed this in the past as a way to store standardized header info, @tacarlson had some objections which I won't go into here. 

An alternative could be to have a master ODD file that has customizations applicable to all modules (such as a master list of language tags) plus where necessary module-specific ODDs that constrain/customize further. My understanding is that ODD files can point to other ODD files as their starting point.

**Comment by nathangibson** [comment]
@tacarlson I'm willing to consider whether including `<langUsage>` in our records is necessary at all. I think we should take into account that the TEI guidelines consider the langUsage info "particularly important if extended language identifiers have been used as the value of xml:lang attributes elsewhere in the document" (http://www.tei-c.org/release/doc/tei-p5-doc/en/html/HD.html#HD41). I think the langUsage could contain a `<p>` describing where to find our language documentation, but including specific languages here is a way to have a certain amount of self-contained documentation packaged with each individual record. But like I said, I'm open to not including it.

Whether or not we use langUsage, we do need a short label and probably description that can be retrieved for each language code we use. For example, we are working on implementing a faceted browse for SBD where you can filter by records containing names in a certain language. We want to display "English" rather then "en", etc., so we need a way to grab a human-readable label using the language code. Its easy to envision other such scenarios. 

I'm open to how this machinable list should relate to the prose documentation, but I do think we need both.

**Comment by tacarlson** [comment]
Thank you for making this clear.  You make a good case for using `<langUsage>` and point to a reasonable use case.  I can see the point that "extended" language identifiers (which I assume means project-specific ones using an "-x-") need to be defined in a place where data consumers can find the definitions.  For the faceted browse feature you mention, one might imagine that it would be enough to build into the operational code the display names of languages and their corresponding language codes, but such operational code would then need to be updated whenever a new language enters our data (such as if we received some name forms in Russian).  So while it is more complex to have some machine-accessible correspondence list between language names and language codes, as long as the relative location of operational code and correspondence list can be maintained, it ought to be more stable as well.

But this raised the further question for me whether our use of "en-x-srp1" is legitimate.  We initially went that route, if I recall correctly, by thinking of this as a kind of simplified transcription system.  Yet in particular, it is not an uncommon case, with data inputs from multiple sources, that the "simplifier" script would simplify one technically transcribed name into a form of the name actually attested in another source.  The simplifier script thinks the language of this name is "en-x-srp1", while the other source thinks the language of this name is simply "en".  This makes me think that the issue is not the language of the name, but its status as an unattested simplification.  Would it make more sense for the simplifier script to generate sources of language "en" and rely on the `@syriaca-tags="#syriaca-simplified-script"` to filter these out of html view?
